# **SMart Supermarket Customer**

In [69]:
# Import Library
import pandas as pd
import numpy as np

## Load Data

In [70]:
#Load Data

df = pd.read_csv("supermarket_ds/Supermarket_Customers.csv", sep="\t")
df

,ID,Year_Birth,Education,Marital_Status,Income,Kidhome,Teenhome,Dt_Customer,Recency,MntWines,...,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Z_CostContact,Z_Revenue,Response
0,5524,1957,Graduation,Single,58138.0,0,0,04-09-2012,58,635,...,7,0,0,0,0,0,0,3,11,1
1,2174,1954,Graduation,Single,46344.0,1,1,08-03-2014,38,11,...,5,0,0,0,0,0,0,3,11,0
2,4141,1965,Graduation,Together,71613.0,0,0,21-08-2013,26,426,...,4,0,0,0,0,0,0,3,11,0
3,6182,1984,Graduation,Together,26646.0,1,0,10-02-2014,26,11,...,6,0,0,0,0,0,0,3,11,0
4,5324,1981,PhD,Married,58293.0,1,0,19-01-2014,94,173,...,5,0,0,0,0,0,0,3,11,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2235,10870,1967,Graduation,Married,61223.0,0,1,13-06-2013,46,709,...,5,0,0,0,0,0,0,3,11,0
2236,4001,1946,PhD,Together,64014.0,2,1,10-06-2014,56,406,...,7,0,0,0,1,0,0,3,11,0
2237,7270,1981,Graduation,Divorced,56981.0,0,0,25-01-2014,91,908,...,6,0,1,0,0,0,0,3,11,0
2238,8235,1956,Master,Together,69245.0,0,1,24-01-2014,8,428,...,3,0,0,0,0,0,0,3,11,0


In [71]:
#result 2240 baris dan 29 kolom
df.shape

(2240, 29)

## Data Inspection

In [72]:
#Dtype: object refer ke universal data spt string 
#untuk Dt_Customer Dtype seharusnya datetime bukan object
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2240 non-null   int64  
 1   Year_Birth           2240 non-null   int64  
 2   Education            2240 non-null   object 
 3   Marital_Status       2240 non-null   object 
 4   Income               2216 non-null   float64
 5   Kidhome              2240 non-null   int64  
 6   Teenhome             2240 non-null   int64  
 7   Dt_Customer          2240 non-null   object 
 8   Recency              2240 non-null   int64  
 9   MntWines             2240 non-null   int64  
 10  MntFruits            2240 non-null   int64  
 11  MntMeatProducts      2240 non-null   int64  
 12  MntFishProducts      2240 non-null   int64  
 13  MntSweetProducts     2240 non-null   int64  
 14  MntGoldProds         2240 non-null   int64  
 15  NumDealsPurchases    2240 non-null   i

In [73]:
# describe - angka
# #hanya dataframe baru menampilkan rangkuman data secara angka krn Dtype nya integer (mean, quartil 1: 25%, median :50%)
df.describe()

,ID,Year_Birth,Income,Kidhome,Teenhome,Recency,MntWines,MntFruits,MntMeatProducts,MntFishProducts,...,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Z_CostContact,Z_Revenue,Response
count,2240.000000,2240.000000,2216.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,...,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.0,2240.0,2240.000000
mean,5592.159821,1968.805804,52247.251354,0.444196,0.506250,49.109375,303.935714,26.302232,166.950000,37.525446,...,5.316518,0.072768,0.074554,0.072768,0.064286,0.013393,0.009375,3.0,11.0,0.149107
std,3246.662198,11.984069,25173.076661,0.538398,0.544538,28.962453,336.597393,39.773434,225.715373,54.628979,...,2.426645,0.259813,0.262728,0.259813,0.245316,0.114976,0.096391,0.0,0.0,0.356274
min,0.000000,1893.000000,1730.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
25%,2828.250000,1959.000000,35303.000000,0.000000,0.000000,24.000000,23.750000,1.000000,16.000000,3.000000,...,3.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
50%,5458.500000,1970.000000,51381.500000,0.000000,0.000000,49.000000,173.500000,8.000000,67.000000,12.000000,...,6.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
75%,8427.750000,1977.000000,68522.000000,1.000000,1.000000,74.000000,504.250000,33.000000,232.000000,50.000000,...,7.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
max,11191.000000,1996.000000,666666.000000,2.000000,2.000000,99.000000,1493.000000,199.000000,1725.000000,259.000000,...,20.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,3.0,11.0,1.000000


we found min in the Year_Birth is 1893 it mean the oldest person was born in 1893. it seems anomaly

In [74]:
df.describe(include='all')

,ID,Year_Birth,Education,Marital_Status,Income,Kidhome,Teenhome,Dt_Customer,Recency,MntWines,...,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Z_CostContact,Z_Revenue,Response
count,2240.000000,2240.000000,2240,2240,2216.000000,2240.000000,2240.000000,2240,2240.000000,2240.000000,...,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.0,2240.0,2240.000000
unique,NaN,NaN,5,8,NaN,NaN,NaN,663,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
top,NaN,NaN,Graduation,Married,NaN,NaN,NaN,31-08-2012,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
freq,NaN,NaN,1127,864,NaN,NaN,NaN,12,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
mean,5592.159821,1968.805804,NaN,NaN,52247.251354,0.444196,0.506250,NaN,49.109375,303.935714,...,5.316518,0.072768,0.074554,0.072768,0.064286,0.013393,0.009375,3.0,11.0,0.149107
std,3246.662198,11.984069,NaN,NaN,25173.076661,0.538398,0.544538,NaN,28.962453,336.597393,...,2.426645,0.259813,0.262728,0.259813,0.245316,0.114976,0.096391,0.0,0.0,0.356274
min,0.000000,1893.000000,NaN,NaN,1730.000000,0.000000,0.000000,NaN,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
25%,2828.250000,1959.000000,NaN,NaN,35303.000000,0.000000,0.000000,NaN,24.000000,23.750000,...,3.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
50%,5458.500000,1970.000000,NaN,NaN,51381.500000,0.000000,0.000000,NaN,49.000000,173.500000,...,6.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
75%,8427.750000,1977.000000,NaN,NaN,68522.000000,1.000000,1.000000,NaN,74.000000,504.250000,...,7.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000


### Check Missing Value

In [75]:
#missing_values = df.isnull().sum().sort_values(ascending=False)
missing_values = df.isna().sum()
missing_values[missing_values > 0]

Income    24
dtype: int64

In [21]:
missing_percentage = round(df.isna().sum()/len(df)*100,2)
missing_percentage

ID                     0.00
Year_Birth             0.00
Education              0.00
Marital_Status         0.00
Income                 1.07
Kidhome                0.00
Teenhome               0.00
Dt_Customer            0.00
Recency                0.00
MntWines               0.00
MntFruits              0.00
MntMeatProducts        0.00
MntFishProducts        0.00
MntSweetProducts       0.00
MntGoldProds           0.00
NumDealsPurchases      0.00
NumWebPurchases        0.00
NumCatalogPurchases    0.00
NumStorePurchases      0.00
NumWebVisitsMonth      0.00
AcceptedCmp3           0.00
AcceptedCmp4           0.00
AcceptedCmp5           0.00
AcceptedCmp1           0.00
AcceptedCmp2           0.00
Complain               0.00
Z_CostContact          0.00
Z_Revenue              0.00
Response               0.00
dtype: float64

the missing value in percentage is only 1.07% so we will remove:

Original dataset: 2,240 customers

Removed missing Income records: 24

Final dataset after Income cleaning: 2,216 customers

### Remove Missing Value

In [76]:
df = df.dropna(subset=["Income"])
df

,ID,Year_Birth,Education,Marital_Status,Income,Kidhome,Teenhome,Dt_Customer,Recency,MntWines,...,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Z_CostContact,Z_Revenue,Response
0,5524,1957,Graduation,Single,58138.0,0,0,04-09-2012,58,635,...,7,0,0,0,0,0,0,3,11,1
1,2174,1954,Graduation,Single,46344.0,1,1,08-03-2014,38,11,...,5,0,0,0,0,0,0,3,11,0
2,4141,1965,Graduation,Together,71613.0,0,0,21-08-2013,26,426,...,4,0,0,0,0,0,0,3,11,0
3,6182,1984,Graduation,Together,26646.0,1,0,10-02-2014,26,11,...,6,0,0,0,0,0,0,3,11,0
4,5324,1981,PhD,Married,58293.0,1,0,19-01-2014,94,173,...,5,0,0,0,0,0,0,3,11,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2235,10870,1967,Graduation,Married,61223.0,0,1,13-06-2013,46,709,...,5,0,0,0,0,0,0,3,11,0
2236,4001,1946,PhD,Together,64014.0,2,1,10-06-2014,56,406,...,7,0,0,0,1,0,0,3,11,0
2237,7270,1981,Graduation,Divorced,56981.0,0,0,25-01-2014,91,908,...,6,0,1,0,0,0,0,3,11,0
2238,8235,1956,Master,Together,69245.0,0,1,24-01-2014,8,428,...,3,0,0,0,0,0,0,3,11,0


In [77]:
df.shape

(2216, 29)

### Check Duplicate Data

In [78]:
duplicate = df.duplicated().sum()
duplicate

np.int64(0)

In [79]:
duplicate_id = df['ID'].duplicated().sum()
duplicate_id

np.int64(0)

### Change Object to datetime for Dt_Customer

In [80]:
df['Dt_Customer'] = pd.to_datetime(df['Dt_Customer'], format='%d-%m-%Y')

# Check result
df[['Dt_Customer']].head()

/var/folders/_v/q8krvj41319240gr34jsthf80000gn/T/ipykernel_14393/4135225054.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Dt_Customer'] = pd.to_datetime(df['Dt_Customer'], format='%d-%m-%Y')


,Dt_Customer
0,2012-09-04
1,2014-03-08
2,2013-08-21
3,2014-02-10
4,2014-01-19


### Check Date Range

In [81]:
#result: df['Dt_Customer'].min(), df['Dt_Customer'].max()
df['Dt_Customer'].min(), df['Dt_Customer'].max()

(Timestamp('2012-07-30 00:00:00'), Timestamp('2014-06-29 00:00:00'))

### Add new Column (Variables)

In [82]:
reference_date = df['Dt_Customer'].max() #dataset terakhir di 2014
reference_date

Timestamp('2014-06-29 00:00:00')

#### Age, Tenure and No of children

In [83]:
# Create Age
df['Age'] = 2014 - df['Year_Birth'] #2014 krn terakhir dataset di 2014

# Create Customer Tenure in days
df['Customer_Tenure'] = (reference_date - df['Dt_Customer']).dt.days

# Create Children
df['Children'] = df['Kidhome'] + df['Teenhome']

df.describe()

/var/folders/_v/q8krvj41319240gr34jsthf80000gn/T/ipykernel_14393/3056654565.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Age'] = 2014 - df['Year_Birth'] #2014 krn terakhir dataset di 2014
/var/folders/_v/q8krvj41319240gr34jsthf80000gn/T/ipykernel_14393/3056654565.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Customer_Tenure'] = (reference_date - df['Dt_Customer']).dt.days
/var/folders/_v/q8krvj41319240gr34jsthf80000gn/T/ipykernel_14393/3056654565.py:8: SettingWithCopyWarning: 
A value i

,ID,Year_Birth,Income,Kidhome,Teenhome,Dt_Customer,Recency,MntWines,MntFruits,MntMeatProducts,...,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Z_CostContact,Z_Revenue,Response,Age,Customer_Tenure,Children
count,2216.000000,2216.000000,2216.000000,2216.000000,2216.000000,2216,2216.000000,2216.000000,2216.000000,2216.000000,...,2216.000000,2216.000000,2216.000000,2216.000000,2216.0,2216.0,2216.000000,2216.000000,2216.000000,2216.000000
mean,5588.353339,1968.820397,52247.251354,0.441787,0.505415,2013-07-10 11:29:27.509025280,49.012635,305.091606,26.356047,166.995939,...,0.073105,0.064079,0.013538,0.009477,3.0,11.0,0.150271,45.179603,353.521209,0.947202
min,0.000000,1893.000000,1730.000000,0.000000,0.000000,2012-07-30 00:00:00,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000,18.000000,0.000000,0.000000
25%,2814.750000,1959.000000,35303.000000,0.000000,0.000000,2013-01-16 00:00:00,24.000000,24.000000,2.000000,16.000000,...,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000,37.000000,180.000000,0.000000
50%,5458.500000,1970.000000,51381.500000,0.000000,0.000000,2013-07-08 12:00:00,49.000000,174.500000,8.000000,68.000000,...,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000,44.000000,355.500000,1.000000
75%,8421.750000,1977.000000,68522.000000,1.000000,1.000000,2013-12-31 00:00:00,74.000000,505.000000,33.000000,232.250000,...,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000,55.000000,529.000000,1.000000
max,11191.000000,1996.000000,666666.000000,2.000000,2.000000,2014-06-29 00:00:00,99.000000,1493.000000,199.000000,1725.000000,...,1.000000,1.000000,1.000000,1.000000,3.0,11.0,1.000000,121.000000,699.000000,3.000000
std,3249.376275,11.985554,25173.076661,0.536896,0.544181,NaN,28.948352,337.327920,39.793917,224.283273,...,0.260367,0.244950,0.115588,0.096907,0.0,0.0,0.357417,11.985554,202.434667,0.749062


In [84]:
df['Age'].describe()

count    2216.000000
mean       45.179603
std        11.985554
min        18.000000
25%        37.000000
50%        44.000000
75%        55.000000
max       121.000000
Name: Age, dtype: float64

Using the IQR method:

IQR = Q3 - Q1

IQR = 55 - 37 = 18

Upper outlier threshold = Q3 + 1.5 × IQR
Upper outlier threshold = 55 + (1.5 × 18)
Upper outlier threshold = 82 

so any customer that has age > 82 can be considered a potential outlier.

we found max age is 121 

In [85]:
# View very old customers
df[df['Age'] > 82][['ID', 'Year_Birth', 'Age']]

,ID,Year_Birth,Age
192,7829,1900,114
239,11004,1893,121
339,1150,1899,115


#### Total Spending

In [86]:
spending_cols = [
    'MntWines',
    'MntFruits',
    'MntMeatProducts',
    'MntFishProducts',
    'MntSweetProducts',
    'MntGoldProds'
]

df['TotalSpend'] = df[spending_cols].sum(axis=1)

/var/folders/_v/q8krvj41319240gr34jsthf80000gn/T/ipykernel_14393/640946077.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['TotalSpend'] = df[spending_cols].sum(axis=1)


#### Total Purchases

In [88]:
purchase_cols = [
    'NumDealsPurchases',
    'NumWebPurchases',
    'NumCatalogPurchases',
    'NumStorePurchases'
]

df['TotalPurchases'] = df[purchase_cols].sum(axis=1)

/var/folders/_v/q8krvj41319240gr34jsthf80000gn/T/ipykernel_14393/3519053257.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['TotalPurchases'] = df[purchase_cols].sum(axis=1)


#### Total Accepted Campaign 

In [89]:
campaign_cols = [
    'AcceptedCmp1',
    'AcceptedCmp2',
    'AcceptedCmp3',
    'AcceptedCmp4',
    'AcceptedCmp5',
    'Response'
]

df['CampaignAcceptedTotal'] = df[campaign_cols].sum(axis=1)

/var/folders/_v/q8krvj41319240gr34jsthf80000gn/T/ipykernel_14393/756571796.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['CampaignAcceptedTotal'] = df[campaign_cols].sum(axis=1)


In [90]:
#Preview
# Preview new variables
df[['Age', 'Customer_Tenure', 'Children', 'TotalSpend', 'TotalPurchases', 'CampaignAcceptedTotal']].head()

,Age,Customer_Tenure,Children,TotalSpend,TotalPurchases,CampaignAcceptedTotal
0,57,663,0,1617,25,1
1,60,113,2,27,6,0
2,49,312,0,776,21,0
3,30,139,1,53,8,0
4,33,161,1,422,19,0


#### Previous Campaign Accepted Total (all campaign with response)

In [114]:
previous_campaign_cols = [
    'AcceptedCmp1',
    'AcceptedCmp2',
    'AcceptedCmp3',
    'AcceptedCmp4',
    'AcceptedCmp5'
]

df['PreviousCampaignAcceptedTotal'] = df[previous_campaign_cols].sum(axis=1)
df[['ID', 
            'PreviousCampaignAcceptedTotal', 
            'CampaignAcceptedTotal']].head()

/var/folders/_v/q8krvj41319240gr34jsthf80000gn/T/ipykernel_14393/315086854.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['PreviousCampaignAcceptedTotal'] = df[previous_campaign_cols].sum(axis=1)


,ID,PreviousCampaignAcceptedTotal,CampaignAcceptedTotal
0,5524,0,1
1,2174,0,0
2,4141,0,0
3,6182,0,0
4,5324,0,0


#### Create Customer Value Segment

In [110]:
# Create customer value segment based on TotalSpend percentile
df['SpendPercentile'] = df['TotalSpend'].rank(pct=True)

def assign_value_segment(p):
    if p >= 0.80:
        return 'High Value'
    elif p >= 0.50:
        return 'Mid Value'
    else:
        return 'Low Value'

df['CustomerValueSegment'] = df['SpendPercentile'].apply(assign_value_segment)

# Check segment distribution
df['CustomerValueSegment'].value_counts(normalize=True) * 100

/var/folders/_v/q8krvj41319240gr34jsthf80000gn/T/ipykernel_14393/2323488474.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['SpendPercentile'] = df['TotalSpend'].rank(pct=True)
/var/folders/_v/q8krvj41319240gr34jsthf80000gn/T/ipykernel_14393/2323488474.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['CustomerValueSegment'] = df['SpendPercentile'].apply(assign_value_segment)


CustomerValueSegment
Low Value     50.000000
Mid Value     30.009025
High Value    19.990975
Name: proportion, dtype: float64

#### Create Retention Segment

In [111]:
# Define thresholds
high_spend_threshold = df['TotalSpend'].quantile(0.80)
medium_spend_threshold = df['TotalSpend'].quantile(0.50)
high_purchase_threshold = df['TotalPurchases'].quantile(0.60)
recency_threshold = 60

def assign_retention_segment(row):
    if row['TotalSpend'] >= high_spend_threshold and row['TotalPurchases'] >= high_purchase_threshold and row['Recency'] < recency_threshold:
        return 'Champions'
    elif row['TotalSpend'] >= high_spend_threshold and row['Recency'] >= recency_threshold:
        return 'At-Risk High Value'
    elif row['TotalSpend'] >= medium_spend_threshold and row['Recency'] < recency_threshold:
        return 'Potential Loyalists'
    elif row['TotalPurchases'] >= high_purchase_threshold:
        return 'Loyal Customers'
    else:
        return 'Low Value'

df['RetentionSegment'] = df.apply(assign_retention_segment, axis=1)

# Check retention segments
df['RetentionSegment'].value_counts()

/var/folders/_v/q8krvj41319240gr34jsthf80000gn/T/ipykernel_14393/1630378729.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['RetentionSegment'] = df.apply(assign_retention_segment, axis=1)


RetentionSegment
Low Value              1119
Potential Loyalists     430
Loyal Customers         251
Champions               232
At-Risk High Value      184
Name: count, dtype: int64

#### Response Comparison

In [116]:
response_comparison = df.groupby('Response').agg(
    CustomerCount=('ID', 'count'),
    AvgIncome=('Income', 'mean'),
    AvgTotalSpend=('TotalSpend', 'mean'),
    AvgTotalPurchases=('TotalPurchases', 'mean'),
    AvgRecency=('Recency', 'mean'),
    AvgChildren=('Children', 'mean'),
    AvgPreviousCampaignAccepted=('PreviousCampaignAcceptedTotal', 'mean')
).round(2)

response_comparison.index = response_comparison.index.map({
    0: 'Non-Responders',
    1: 'Responders'
})

response_comparison

,CustomerCount,AvgIncome,AvgTotalSpend,AvgTotalPurchases,AvgRecency,AvgChildren,AvgPreviousCampaignAccepted
Response,,,,,,,
Non-Responders,1883,50839.13,540.12,14.38,51.44,1.00,0.18
Responders,333,60209.68,985.66,17.69,35.26,0.65,0.99


### Remove Outliers 

#### Age

In [117]:
# Remove age outliers
data_clean = df[df['Age'] <= 82].copy()

data_clean.shape

(2213, 39)

#### Income

In [118]:
data_clean['Income'].describe()

count      2213.000000
mean      52236.581563
std       25178.603047
min        1730.000000
25%       35246.000000
50%       51373.000000
75%       68487.000000
max      666666.000000
Name: Income, dtype: float64

Using the IQR method:

IQR = Q3 - Q1

IQR = 68,487 - 35,246

IQR = 33,241

Upper limit = Q3 + 1.5 × IQR

Upper limit = 118,348.5

In [119]:
data_clean = data_clean[data_clean['Income'] < 118348].copy()
data_clean.shape
#result: (2205,35)

(2205, 39)

#### Z_CostContact and Z_Revenue

In [120]:
df[["Z_CostContact", "Z_Revenue"]].describe()

,Z_CostContact,Z_Revenue
count,2216.0,2216.0
mean,3.0,11.0
std,0.0,0.0
min,3.0,11.0
25%,3.0,11.0
50%,3.0,11.0
75%,3.0,11.0
max,3.0,11.0


In [121]:
df[["Z_CostContact", "Z_Revenue"]].nunique()

Z_CostContact    1
Z_Revenue        1
dtype: int64

 Z_CostContact dan Z_Revenue always the same for most customers so dont provide useful info. 

In [122]:
# Drop  columns
data_clean = data_clean.drop(columns=['Z_CostContact', 'Z_Revenue'])

### Final Data Quality Check

In [123]:
# Final missing value check
data_clean.isnull().sum().sort_values(ascending=False).head(10)

ID                   0
NumWebVisitsMonth    0
AcceptedCmp4         0
AcceptedCmp5         0
AcceptedCmp1         0
AcceptedCmp2         0
Complain             0
Response             0
Age                  0
Customer_Tenure      0
dtype: int64

In [124]:
# Final shape
data_clean.shape

(2205, 37)

In [125]:
# Final columns
data_clean.columns

Index(['ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome',
       'Teenhome', 'Dt_Customer', 'Recency', 'MntWines', 'MntFruits',
       'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts',
       'MntGoldProds', 'NumDealsPurchases', 'NumWebPurchases',
       'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth',
       'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1',
       'AcceptedCmp2', 'Complain', 'Response', 'Age', 'Customer_Tenure',
       'Children', 'TotalSpend', 'TotalPurchases', 'CampaignAcceptedTotal',
       'SpendPercentile', 'CustomerValueSegment', 'RetentionSegment',
       'PreviousCampaignAcceptedTotal'],
      dtype='object')

### Reorder Column

In [126]:
data_clean = data_clean[
    [
        'ID', 'Year_Birth', 'Age', 'Education', 'Marital_Status', 'Income',
        'Kidhome', 'Teenhome', 'Children',
        'Dt_Customer', 'Customer_Tenure', 'Recency',
        'MntWines', 'MntFruits', 'MntMeatProducts', 'MntFishProducts',
        'MntSweetProducts', 'MntGoldProds', 'TotalSpend',
        'NumDealsPurchases', 'NumWebPurchases', 'NumCatalogPurchases',
        'NumStorePurchases', 'NumWebVisitsMonth', 'TotalPurchases',
        'AcceptedCmp1', 'AcceptedCmp2', 'AcceptedCmp3', 'AcceptedCmp4',
        'AcceptedCmp5', 'Response', 'PreviousCampaignAcceptedTotal',
        'CampaignAcceptedTotal',
        'Complain',
        'SpendPercentile', 'CustomerValueSegment', 'RetentionSegment'
    ]
]

### Export Clean Dataset

In [127]:
# Export cleaned dataset
data_clean.to_csv("supermarket_customers_cleaned.csv", index=False)

print("Cleaned dataset exported successfully.")

Cleaned dataset exported successfully.
